src/pkg/artifact/dao/dao_test.go

```diff
@@ -0,0 +1,201 @@
+	suite.Suite
+	require    *require.Assertions
+	assert     *assert.Assertions
+	dao        DAO
+	artifactID int64
+}
+
+func (d *daoTestSuite) SetupSuite() {
+	d.dao = New()
+	d.require = require.New(d.T())
+	d.assert = assert.New(d.T())
+	common_dao.PrepareTestForPostgresSQL()
+}
+
+func (d *daoTestSuite) SetupTest() {
+	artifact := &Artifact{
+		Type:              typee,
+		MediaType:         mediaType,
+		ManifestMediaType: manifestMediaType,
+		ProjectID:         projectID,
+		RepositoryID:      repositoryID,
+		Digest:            digest,
+		Size:              1024,
+		PushTime:          time.Now(),
+		PullTime:          time.Now(),
+		ExtraAttrs:        `{"attr1":"value1"}`,
+		Annotations:       `{"anno1":"value1"}`,
+	}
+	id, err := d.dao.Create(nil, artifact)
+	d.require.Nil(err)
+	d.artifactID = id
+}
+
+func (d *daoTestSuite) TearDownTest() {
+	err := d.dao.Delete(nil, d.artifactID)
+	d.require.Nil(err)
+}
+
+func (d *daoTestSuite) TestGetTotal() {
+	// nil query
+	total, err := d.dao.Count(nil, nil)
+	d.require.Nil(err)
+	d.assert.True(total > 0)
+	// query by repository ID and digest
+	total, err = d.dao.Count(nil, &q.Query{
+		Keywords: map[string]interface{}{
+			"repository_id": repositoryID,
+			"digest":        digest,
+		},
+	})
+	d.require.Nil(err)
+	d.assert.Equal(int64(1), total)
+}
+
+func (d *daoTestSuite) TestList() {
+	// nil query
+	artifacts, err := d.dao.List(nil, nil)
+	d.require.Nil(err)
+	found := false
+	for _, artifact := range artifacts {
+		if artifact.ID == d.artifactID {
+			found = true
+			break
+		}
+	}
+	d.assert.True(found)
+
+	// query by repository ID and digest
+	artifacts, err = d.dao.List(nil, &q.Query{
+		Keywords: map[string]interface{}{
+			"repository_id": repositoryID,
+			"digest":        digest,
+		},
+	})
+	d.require.Nil(err)
+	d.require.Equal(1, len(artifacts))
+	d.assert.Equal(d.artifactID, artifacts[0].ID)
+}
+
+func (d *daoTestSuite) TestGet() {
+	// get the non-exist artifact
+	_, err := d.dao.Get(nil, 10000)
+	d.require.NotNil(err)
+	d.assert.True(ierror.IsErr(err, ierror.NotFoundCode))
+
+	// get the exist artifact
+	artifact, err := d.dao.Get(nil, d.artifactID)
+	d.require.Nil(err)
+	d.require.NotNil(artifact)
+	d.assert.Equal(d.artifactID, artifact.ID)
+}
+
+func (d *daoTestSuite) TestCreate() {
+	// the happy pass case is covered in Setup
+
+	// conflict
+	artifact := &Artifact{
+		Type:              typee,
+		MediaType:         mediaType,
+		ManifestMediaType: manifestMediaType,
+		ProjectID:         projectID,
+		RepositoryID:      repositoryID,
+		Digest:            digest,
+		Size:              1024,
+		PushTime:          time.Now(),
+		PullTime:          time.Now(),
+		ExtraAttrs:        `{"attr1":"value1"}`,
+		Annotations:       `{"anno1":"value1"}`,
+	}
+	_, err := d.dao.Create(nil, artifact)
+	d.require.NotNil(err)
+	d.assert.True(ierror.IsErr(err, ierror.ConflictCode))
+}
+
+// Delete is covered in TearDown
+
+func (d *daoTestSuite) TestUpdate() {
+	// pass
```

Comment:
should test the case when the artifact to be updated does not exist.

Reply:
Done